File: java/cours/src/session1/td/SuccRec.java
package session1.td;

public class SuccRec extends Succ implements Nat {

	public static FabriqueNaturels<Nat> FAB = new SuccRec(NatParInt.FAB.creerNatAvecValeur(0));
	
	public SuccRec(Nat predecesseur) {
		super(predecesseur);
	}
	
	//Create the successor of an object
	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return FAB.creerNatAvecValeur(predecesseur.val()+1);
	}

	//Add two numbers (the pointed one and the one passed in arg)
	@Override
	public Nat somme(Nat x) {
		return FAB.creerNatAvecValeur(this.val() + x.val());
	}

	//Return a neutral element of the addition, a.k.a zero
	@Override
	public Nat zero() {
		return new Zero();
	}

	//Multiplication between two numbers
	@Override
	public Nat produit(Nat x) {
		return FAB.creerNatAvecValeur(this.val()*x.val());
	}

	//Return a neutral element of the multiplication, a.k.a one
	@Override
	public Nat un() {
		return FAB.creerNatAvecValeur(1);
	}

	//Return the rest of the Euclidean division this.val/x
	@Override
	public Nat modulo(Nat x) {
		return (x.estNul()) ? null : FAB.creerNatAvecValeur(this.val()%x.val());
	}

	//Return the quotient of the division this.val/x
	@Override
	public Nat div(Nat x) {
		return (x.estNul()) ? null : FAB.creerNatAvecValeur(this.val()/x.val());
	}

	//Equals method
	public boolean equals(Object o) {
		if(!(o instanceof Nat)) return false;
		Nat n = (Nat)o;
		return this.val() == n.val();
	}

}

File: typescript/succRec.ts
import { Nat } from "./Nat";
import { FabriqueNaturels } from "./FabriqueNaturels";
import { NatParInt } from "./NatParInt";
import { Succ } from "./Succ";
import { Zero } from "./Zero";

export class SuccRec extends Succ implements Nat {

    public static FAB: FabriqueNaturels<Nat> = new SuccRec(NatParInt.FAB.creerNatAvecValeur(0));

    public constructor (pred: Nat){
        super(pred);
    }
    estNul(): boolean {
        return this.val() == 0;
    }
    chiffre(i: number): number {
        if(i < 0 && i > this.val().toString.length)
            throw new Error("Invalid argument");

        var number = this.val().toString();
        var length = number.length;
        i = length - i;

        return parseInt(number.charAt(i));
    }
    taille(): number {
        return this.val().toString().length;
    }
    equals(n: Nat): boolean {
        return this.val() == n.val();
    }
    
    creerZero(): Nat {
        return Zero.FAB.creerZero();
    }
    creerSuccesseur(predecesseur: Nat): Nat {
        return new SuccRec(predecesseur);
    }
   
    somme(x: Nat): Nat {
        if(x.estNul())
            return this;
        return this.creerSuccesseur(this).somme(x.predecesseur())
    }

    produit(x: Nat): Nat {
        return this.creerNatAvecValeur(this.val() * x.val());
    }

    modulo(x: Nat): Nat {
        if (x.estNul()) 
            throw new Error("Division par 0");
        return this.creerNatAvecValeur(this.val() % x.val());
    }
    div(x: Nat): Nat {
        if (x.estNul()) 
            throw new Error("Division par 0");
        return this.creerNatAvecValeur(this.val() / x.val());
    }
    toString(): String {
        return String(this.val());
    }

}